### PR TITLE
Update lr_logger.py

### PR DESCRIPTION
**Pull request by aiyolo**
when logging learning_rate, we should provide `step` choices, 
for example, we might want to call a `lr_logger` with `global_step` or `current_epoch`, now we can do that like
`lr_logger = LearningRateLogger(global_step=[False| True])`
`trainer = pl.Trainer(callbacks=[lr_logger])`

## What does this PR do?

Fixes # (issue)

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/lr_logger.py [hunk 1]
```diff
@@ -67,13 +68,15 @@ def on_train_start(self, trainer, pl_module):
     def on_train_batch_start(self, trainer, pl_module):
         latest_stat = self._extract_lr(trainer, 'step')
         if trainer.logger and latest_stat:
-            trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+            step = trainer.global_step if self._global_step == True else trainer.current_epoch
```
Comment (justusschock) [inline pytorch_lightning/callbacks/lr_logger.py:71]: can we outsource this to a property to avoid duplication?

File: pytorch_lightning/callbacks/lr_logger.py [hunk 2]
```diff
@@ -38,9 +38,10 @@ def configure_optimizer(self):
                             'name': 'my_logging_name'}
             return [optimizer], [lr_scheduler]
     """
-    def __init__(self):
+    def __init__(self, global_step=True):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/lr_logger.py:41]: can you add some type annotations and docstrings here?

Comment (aiyolo) [reply pytorch_lightning/callbacks/lr_logger.py:41]: yes, I will have a try

**Comment by Borda** [comment]
@aiyolo mind mark comments as resolved if you did it so we know it is ready for another review round...

**Comment by rohitgr7** [comment]
Just a suggestion: How about `interval` or `step` parameter that takes a string `'global_step' or 'epoch'` just like we have for schedulers :) ?

**Comment by aiyolo** [comment]
CI test fails with errors like `assert all(len(lr) == trainer.max_epochs for k, lr in lr_logger.lrs.items()), \
            'Length of logged learning rates exceeds the number of epochs'`

I think the reason is that,  in current implementation, we record learning_rate on every epoch, and the x-axis is global_step, so that `len(lr) == trainer.max_epochs`, however, in my commits, we record learnning_rate every step in default, therefore, CI test fails.

so, how to deal with this confict? to modify the  test case or my my commits?

**Comment by SkafteNicki** [comment]
@aiyolo if you change:
`all(len(lr) == trainer.max_epoch for k, lr in lr_logger.lrs.items())`
to
`all(len(lr) == trainer.global_step for k, lr in lr_logger.lrs.items())`
the tests should pass

**Comment by SkafteNicki** [comment]
On second thought, it is probably better if we parametrize this so we test for both settings. Something like this should do the job:

```
@pytest.mark.parametrize("logging_interval", ['step', 'epoch'])
def test_lr_logger_multi_lrs(tmpdir, logging_interval):
    """ Test that learning rates are extracted and logged for multi lr schedulers. """
    tutils.reset_seed()

    model = EvalModelTemplate()
    model.configure_optimizers = model.configure_optimizers__multiple_schedulers

    lr_logger = LearningRateLogger(interval=logging_interval)
    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=2,
        limit_val_batches=0.1,
        limit_train_batches=0.5,
        callbacks=[lr_logger],
    )
    result = trainer.fit(model)
    assert result

    assert lr_logger.lrs, 'No learning rates logged'
    assert len(lr_logger.lrs) == len(trainer.lr_schedulers), \
        'Number of learning rates logged does not match number of lr schedulers'
    assert all([k in ['lr-Adam', 'lr-Adam-1'] for k in lr_logger.lrs.keys()]), \
        'Names of learning rates not set correctly'
    if logging_interval=='step':
        expected_number_logged = trainer.global_step
    else:
        expected_number_logged = trainer.current_epoch+1 # zero based
    assert all(len(lr) == expected_number_logged for k, lr in lr_logger.lrs.items()), \
        'Length of logged learning rates do not match the expected number'

```

**Comment by SkafteNicki** [comment]
@aiyolo what will happen if I have one scheduler that updates every epoch and one scheduler that updates every step? will the results will be logged correctly?

**Comment by aiyolo** [comment]
> @aiyolo what will happen if I have one scheduler that updates every epoch and one scheduler that updates every step? will the results will be logged correctly?

`lr_logger.py` first extracts `lr` from an optimizer and then log it, so if scheduler updates every epoch(100 batch_step per epoch), and we log lr every step, then we will record same lrs in  the 100 batch step.  For detailed results, please refer to the images below.

for example, if we use 
```
schedulers = [{'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizers[0], 0.99), 
    'interval': 'epoch'}]
lr_logger = LearningRateLogger(interval='step')
```
cheduler_step_vs_interval_step
scheduler_step_vs_interval_step
scheduler_step_vs_interval_epoc
scheduler_step_vs_interval_epoch
scheduler_epoch_vs_interval_step
scheduler_epoch_vs_interval_step
scheduler_epoch_vs_interval_epoch
scheduler_epoch_vs_interval_epoch

**Comment by SkafteNicki** [comment]
@aiyolo so maybe my question was not clear enough: if i have two scheduler defined as 
```
scheduler1 = {'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizers[0], 0.99), 
    'interval': 'epoch'}
scheduler2 = {'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizers[1], 0.99), 
    'interval': 'step'}
schedulers = [scheduler1, scheduler2]
```
can I still have the option to log scheduler1 every epoch and scheduler2 every step (match how often they update) or will this force them to log at the same time?

**Comment by aiyolo** [comment]
> @aiyolo so maybe my question was not clear enough: if i have two scheduler defined as
> 
> ```
> scheduler1 = {'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizers[0], 0.99), 
>     'interval': 'epoch'}
> scheduler2 = {'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizers[1], 0.99), 
>     'interval': 'step'}
> schedulers = [scheduler1, scheduler2]
> ```
> 
> can I still have the option to log scheduler1 every epoch and scheduler2 every step (match how often they update) or will this force them to log at the same time?

yes, this will force them to log at the time. to support different log_interval, I think we should add an optional key in every scheduler, like {'sheduler':..., 'log_interval':...}.

**Comment by SkafteNicki** [comment]
But the current implementation support this already. The log interval is determined by the `"interval" ` key.

**Comment by aiyolo** [comment]
emm..., so you mean we should log `lr` according to `interval` key in the scheduler, however, in this case, what if we want to record `lr` every epoch, but the `interval='step'` in the scheduler?

**Comment by SkafteNicki** [comment]
Then I guess the compromise is:
LrLoggerCallback(interval=None): default behavior, lr are loggede according to their individual "interval" key
LrLoggerCallback(interval="epoch"): force all lr to log every epoch
LrLoggerCallback(interval="step): force all lr to log every step

**Comment by aiyolo** [comment]
much better!

File: pytorch_lightning/callbacks/lr_logger.py [hunk 3]
```diff
@@ -64,22 +70,36 @@ def on_train_start(self, trainer, pl_module):
...
+        if self.logging_interval is None:
+            latest_stat = self._extract_lr(trainer, 'step')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+        elif self.logging_interval == 'step':
+            latest_stat = self._extract_lr(trainer, 'any')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+        else:
+            pass
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lr_logger.py:83]:
```suggestion
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 4]
```diff
@@ -64,22 +70,36 @@ def on_train_start(self, trainer, pl_module):
...
+        if self.logging_interval is None:
+            latest_stat = self._extract_lr(trainer, 'epoch')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.current_epoch)
+        elif self.logging_interval == 'epoch':
+            latest_stat = self._extract_lr(trainer, 'any')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.current_epoch)
+        else:
+            pass
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lr_logger.py:95]:
```suggestion
```
How about having a MisconfigurationException if `logging_interval` is not set to either None, 'step' or 'epoch'?

File: pytorch_lightning/callbacks/lr_logger.py [hunk 5]
```diff
@@ -80,7 +80,8 @@ def on_batch_start(self, trainer, pl_module):
             if trainer.logger and latest_stat:
                 trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
         else:
-            pass
+            raise MisconfigurationException(
+                'logging_interval should be `step` or `epoch` or None.')
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lr_logger.py]: @aiyolo I would prefer this check in `__init__` itself, so that one should know this while creating the callback rather than after doing `.fit` and remove the else part completely :)

Comment (aiyolo) [reply pytorch_lightning/callbacks/lr_logger.py]: thanks for your suggestion

File: pytorch_lightning/callbacks/lr_logger.py [hunk 6]
```diff
@@ -63,22 +77,33 @@ def on_train_start(self, trainer, pl_module):
...
-            trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+    def on_batch_start(self, trainer, pl_module):
+        if self.logging_interval is None:
+            latest_stat = self._extract_lr(trainer, 'step')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+        elif self.logging_interval == 'step':
+            latest_stat = self._extract_lr(trainer, 'any')
+            if trainer.logger and latest_stat:
+                trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:88]: double checking, is it really None -> step and step -> any?

File: tests/callbacks/test_lr_logger.py [hunk 7]
```diff
@@ -50,14 +50,15 @@ def test_lr_logger_no_lr(tmpdir):
         assert result
 
 
-def test_lr_logger_multi_lrs(tmpdir):
+@pytest.mark.parametrize("logging_interval", ['step', 'epoch'])
+def test_lr_logger_multi_lrs(tmpdir, logging_interval):
```
Comment (awaelchli) [inline tests/callbacks/test_lr_logger.py:54]: is `logging_interval=None` tested elsewhere?

File: pytorch_lightning/callbacks/lr_logger.py (hunk 6 above)
Comment (aiyolo) [reply pytorch_lightning/callbacks/lr_logger.py:88]: if `logging_interval` is set to` None`, then only `lr` that updates every step will be extracted,` lr` that updates every epoch will be skipped, finally we log extracted lr every step,
if set to `'step'`, lr that updates every step and every epoch will be both extracted, finally we log` lr` every step.
similarly for `epoch`.

Comment (aiyolo) [reply pytorch_lightning/callbacks/lr_logger.py:88]: see https://github.com/PyTorchLightning/pytorch-lightning/pull/2847#issuecomment-670557150

File: tests/callbacks/test_lr_logger.py (hunk 7 above)
Comment (aiyolo) [reply tests/callbacks/test_lr_logger.py:54]: not yet, if `logging_interval=None`, then the assertion `assert all(len(lr) == expected_number_logged for k, lr in lr_logger.lrs.items())` will depend on individual scheduler

File: pytorch_lightning/callbacks/lr_logger.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/lr_logger.py:88]: Ok, thanks for explaining. I think I understand. I guess the confusing part here was that I was just looking at on_batch_start but the other half of the logging is happening in on_epoch_start. 
Nonetheless I have a suggestion regarding the code. The way I would have implemented it is defining a mapping
```python
mapping = {None: "step", "step": any} 
```
and then get rid of the duplicated code in the if blocks and just do:
```python
extract_key = mapping[self.logging_interval]
latest_stat = self._extract_lr(trainer, extract_key)
            if trainer.logger and latest_stat:
                trainer.logger.log_metrics(latest_stat, step=trainer.global_step)

```
and analogous in epoch_end method. What do you think of this?

Comment (aiyolo) [reply pytorch_lightning/callbacks/lr_logger.py:88]: It looks better, could you commit the changes.

Comment (awaelchli) [reply pytorch_lightning/callbacks/lr_logger.py:88]: sorry no, I don't have access to your branch to push changes.

Comment (rohitgr7) [reply pytorch_lightning/callbacks/lr_logger.py:88]: I don't think this will work because `on_batch_start` will be called even when logging_interval == 'epoch'. let me try something else.

Comment (awaelchli) [reply pytorch_lightning/callbacks/lr_logger.py]: does your reply refer to my suggestion or the initial question I had in this thread?

Comment (rohitgr7) [reply pytorch_lightning/callbacks/lr_logger.py]: your suggestion

Comment (rohitgr7) [reply pytorch_lightning/callbacks/lr_logger.py]: although I realized later that your suggestion will work with an update. made those changes :)

Comment (awaelchli) [reply pytorch_lightning/callbacks/lr_logger.py]: ok, maybe I made a typo somewhere, but the idea was to make it a refactor, so equivalent code. 
anyway I think it is strange that you had to make a change and the tests did not fail? Doesn't that mean the test is incomplete?

Comment (awaelchli) [reply pytorch_lightning/callbacks/lr_logger.py]: ok my bad, not enough sleep 😅
